OCPVE-672: feat: add the OLM as a capability

eggfoobar opened the pull request:
Add the OLM as a capability.

Note, that OLM and Marketplace have a dependency, the Marketplace should not be enabled and the OLM disabled. This enforcement will be done at the installer level similar to how we enforce [MachineAPI](https://github.com/openshift/installer/blob/7f49fee68faee3f4def8e0cbbf4c27efc3488859/pkg/types/validation/installconfig.go#L162)

eggfoobar commented:
/assign @JoelSpeed 
/cc @kevinrizza

JoelSpeed commented:
@eggfoobar Who can LGTM this from the OLM team to say they're happy that this will be completed in 4.15?

kevinrizza commented:
/lgtm

JoelSpeed commented on config/v1/types_cluster_version.go:
```diff
@@ -267,6 +267,9 @@ const (
 	// ClusterVersionCapabilityMarketplace manages the Marketplace operator which
 	// supplies Operator Lifecycle Manager (OLM) users with default catalogs of
 	// "optional" operators.
+	//
+	// Note that Marketplace has a hard requirement on OLM. OLM can not be disabled
+	// while Marketplace is enabled.
```
> +1, however my understanding is that we can include marketplace manifests in the OLM capability.

I thought that would be the case for baremetal and machine api, but, it doesn't work like that. There is no `or` semantic for capabailities, so the only way to do make sure marketplace is enabled correctly or not is with API validation, eg https://github.com/openshift/api/pull/1537

@eggfoobar Can you add validation similar to https://github.com/openshift/api/pull/1537 to this PR?

eggfoobar replied:
I'm not 100% sure if we can do that enforcement via the manifests, @wking Would you know if that's possible? Right now I have a WIP PR that adds the logic in the installer https://github.com/openshift/installer/pull/7495/files#diff-3f811ba555e4ce01f184d351af28272069d04129a14be43b6ec1823be8b38221R166-R169

joelanford replied:
The component selection enhancement seems to pretty clearly indicate that marketplace can list `capability.openshift.io/name=marketplace+OLM` and get the behavior I describe. 

> If a resource participates in multiple capabilities, it should specify all capabilities as plus-sign separated values in the annotation(e.g. capability.openshift.io/name=console+monitoring). The resource will only be included if all the capabilities on the resource are enabled. This allows a component to define a ServiceMonitor which is part of both the monitoring capability and the Foo capability. If either of those capabilities are excluded, the ServiceMonitor should not be created. (If monitoring is excluded then no ServiceMonitors should be created, if only Foo is excluded, then only Foo's ServiceMonitors should be excluded while ServiceMonitors associated with other capabilities would not be).

Is this a gap or a bug in the component selection feature?

eggfoobar replied:
Thanks @joelanford, not sure how I missed that snippet. I'll update this PR to include the validation you mentioned @JoelSpeed , thanks folks!

eggfoobar replied:
Updated with changes

wking replied:
If folks are ok with this sort of dependency approach going forward, it's probably worth updating [this section][1] of the enhancement and adding a new header and subsection for dependencies being grudgingly acceptable with the implementation plan being API-side validations and installer guard logic, with links to specific pulls that folks think set good precedent for each.

[1]: https://github.com/openshift/enhancements/blob/286e46ad8c1fc0baf9f940b21ea85381c3e54640/enhancements/installer/component-selection.md?plain=1#L349-L357

eggfoobar commented:
@joelanford and @JoelSpeed This should be good for another look, I went ahead and updated the annotation to include `+OLM` and added the validation and e2e test.

eggfoobar commented:
## Validation

This feature was validated on a cluster by enabling everything except for `OLM` and `marketplace`. During testing a few operators had some issues either expecting OLM resources to be there or applying OLM resources. PRs were opened up to resolve those errors and after building a new payload the cluster install succeeded and things seemed to be operational.

### Fixes Needed

| Operator | Reason | PR |
|----------|---------|----|
| Node Tuning | needed extra error handling for no resource existing | openshift/cluster-node-tuning-operator#801 |
| Cluster Monitoring | needed added OLM annotation on an OLM resource in it's manifests |  openshift/cluster-monitoring-operator#2095 |
| Console Operator | requires input from console team since I'm missing some context on usage of OLM but otherwise, needed query for EnabledCapabilities to not look for OLM resource during healthcheck | openshift/console-operator#795 |

### Cluster Configuration

**Platform:** 
 `aws`
**ClusterBot Payload:**
`build 4.15,openshift/cluster-config-operator#357,openshift/cluster-version-operator#971,openshift/operator-framework-olm#565,openshift/cluster-node-tuning-operator#801,openshift/cluster-monitoring-operator#2095,openshift/console-operator#795`
**CapabilitiesConfig:**
```yaml
capabilities:
  baselineCapabilitySet: None
  additionalEnabledCapabilities:
  - openshift-samples
  - baremetal
  - Console
  - Insights
  - Storage
  - CSISnapshot
  - NodeTuning
  - MachineAPI
  - Build
  - DeploymentConfig
  - ImageRegistry
```

JoelSpeed commented on config/v1/0000_03_marketplace-operator_01_operatorhub.crd.yaml:
```diff
@@ -3,7 +3,7 @@ kind: CustomResourceDefinition
   annotations:
     api-approved.openshift.io: https://github.com/openshift/api/pull/470
-    capability.openshift.io/name: marketplace
+    capability.openshift.io/name: marketplace+OLM
```
Does this need to be changed if we are ensuring that OLM is always enabled when a cluster has marketplace enabled?

What is the effect of this on a cluster that upgrades into 4.15?

eggfoobar replied:
After testing it, that shouldn't be needed, I assumed given enhancement docs we might want to have it to make explicit in the capability name that a dependency exists, but this change didn't even seem to alter the payload CR anyway. I went ahead and removed it. Regarding the upgrades into 4.15, I added an upgrade validation comment. That behaved as expected as well, meaning the `ClusterVersion` was updated and the events showed the enforcement logic working.

JoelSpeed commented on config/v1/stable.clusterversion.testsuite.yaml:
```diff
@@ -130,6 +130,38 @@ tests:
+          additionalEnabledCapabilities:
+          - marketplace
+          - OLM
+  - name: Should not be able to create a ClusterVersion with base capability None, and additional capabilities baremetal without MachineAPI
```
Name is wrong here

eggfoobar replied:
oops, fixed

eggfoobar commented:
# Upgrade Validation
I was able to validate upgrading form a 4.14 cluster with only `marketplace` and `MachineAPI`(needed for aws platform) to a 4.15 cluster with the new optional OLM feature. The upgrade was done via the `openshift-tests run-upgrade none --to-image=registry.build05.ci.openshift.org/ci-ln-g4wlytt/release:latest`. The results show that `enabledCapabilities` correctly got updated to include `OLM` since marketplace was enabled.

<table>
<tr>
<td> Initial ClusterVersion State </td> <td> After Upgrade ClusterVersion State </td>
</tr>
<tr>
<td>

```yaml
apiVersion: config.openshift.io/v1
kind: ClusterVersion
metadata:
  name: version
spec:
  capabilities:
    additionalEnabledCapabilities:
    - marketplace
    - MachineAPI
    baselineCapabilitySet: None
  channel: stable-4.14
status:
  capabilities:
    enabledCapabilities:
    - MachineAPI
    - marketplace
    knownCapabilities:
    - Build
    - CSISnapshot
    - Console
    - DeploymentConfig
    - ImageRegistry
    - Insights
    - MachineAPI
    - NodeTuning
    - Storage
    - baremetal
    - marketplace
    - openshift-samples 
``` 
</td>
<td>

```yaml
apiVersion: config.openshift.io/v1
kind: ClusterVersion
metadata:
  name: version
spec:
  capabilities:
    additionalEnabledCapabilities:
    - marketplace
    - MachineAPI
    baselineCapabilitySet: None
  channel: stable-4.14
status:
  capabilities:
    enabledCapabilities:
    - MachineAPI
    - OLM
    - marketplace
    knownCapabilities:
    - Build
    - CSISnapshot
    - Console
    - DeploymentConfig
    - ImageRegistry
    - Insights
    - MachineAPI
    - NodeTuning
    - OLM
    - Storage
    - baremetal
    - marketplace
    - openshift-samples 
```
</td>
</tr>
</table>

### Cluster Configuration

**Platform:** 
 `aws`

**Payload:**
from: 
`quay.io/openshift-release-dev/ocp-release:4.14.0-rc.2-x86_64`
to (image built with ClusterBot):
`build 4.15,openshift/cluster-config-operator#357,openshift/cluster-version-operator#971,openshift/operator-framework-olm#565,openshift/cluster-node-tuning-operator#801,openshift/cluster-monitoring-operator#2095,openshift/console-operator#795`

**Install Capabilities Config:**
```yaml
capabilities:
  baselineCapabilitySet: None
  additionalEnabledCapabilities:
  - "marketplace"
  - "MachineAPI"
```

wking commented on config/v1/types_cluster_version.go:
```diff
@@ -331,9 +335,14 @@ const (
+
+	// ClusterVersionCapabilityOLM manages the Operator Lifecycle Manager (OLM)
+	// which itself manages the lifecycle of operators
+	ClusterVersionCapabilityOLM ClusterVersionCapability = "OLM"
```
Nit: everyone on this pull request knows what OLM unpacks to, but to make life easier for new OpenShift users/customers, can we unpack this to `OperatorLifecycleManager`?

eggfoobar replied:
I like that, changing to full name to remove ambiguity. Thanks!

kevinrizza commented:
/lgtm

JoelSpeed commented:
/lgtm
